Split histogram classes and into JSON and binary classes

idoqo opened the pull request:
Ongoing patch for milestone #4, this splits up the histogram base class into `Histogram_binary` and `Histogram_json`. A question though:
- The code that differs between `Histogram_builder_json` and `Histogram_builder` isn't much (just the `next` implementation), does it make sense to unify them and just reference `Histogram_base`?

Also, the code builds and seems to work fine with building JSON from histograms (by inspecting `mysqld.1.err` at least), but it messes up storing the JSON structure in the table - I'm not sure why that's happening at the moment.

idoqo committed:
Split histogram classes and into JSON and binary classes

Signed-off-by: Michael Okoko <okokomichaels@outlook.com>

spetrunia commented on sql/sql_statistics.h:
```diff
@@ -153,6 +153,24 @@ class Histogram_base : public Sql_alloc
+
+  virtual void init_for_collection(MEM_ROOT *mem_root, Histogram_type htype_arg, ulonglong size)=0;
+
+  virtual bool is_available()=0;
```
Need to look at histogram::is_available(). 

This function made sense in the old code, where it was possible that Histogram object is present but is "empty".  This should no longer be the case. (I am not sure whether this belongs to this milestone.. just making a note about something to think about)

spetrunia commented on sql/sql_statistics.h:
```diff
@@ -326,9 +339,54 @@ class Histogram : public Histogram_base
+  uchar *values;
+
+public:
+  bool parse(MEM_ROOT *mem_root, Histogram_type type_arg, const uchar *ptr, uint size) override {return false;}
```
Why is this function empty? Don't we have the code already that parses the histogram from its JSON implementation?

spetrunia commented on sql/sql_statistics.h:
```diff
@@ -326,9 +339,54 @@ class Histogram : public Histogram_base
+
+  void serialize(Field *to_field) override{}
+
+  uint get_size() override {return (uint) size;}
```
What does this function do?  Number of buckets is returned by `get_width` (see below). Please check if we can remove it.

idoqo replied:
I think it's used currently to fill the `hist_size` column for binary histograms. We can use `get_width` for that, but I think it gets halved for double-precision-binary-histograms.

spetrunia commented on sql/sql_statistics.cc:
```diff
@@ -2608,18 +2641,18 @@ bool Column_statistics_collected::add()
   case JSON:
-    return new Histogram();
+    return new Histogram_binary();
   default:
```
Why is Histogram_binary created when type=JSON? 
.. also I see that the function is not used anywhere...

idoqo committed:
remove unneeded shared methods

Signed-off-by: Michael Okoko <okokomichaels@outlook.com>

spetrunia commented on sql/sql_statistics.cc:
```diff
@@ -1086,7 +1086,7 @@ class Column_stat: public Stat_table
           stat_field->store(stats->histogram_? 
-                              stats->histogram_->get_size() : 0);
+                              stats->histogram_->get_width() : 0);
           break;
```
Do I read it correctly: this function now stores the number of buckets, not the number of bytes?
Please don't do this as this breaks compatibility with older code.

Instead, please check the length of the data that was written by `stats->histogram_->serialize(stat_field);` and store that (yes, this means you need to process COLUMN_STAT_HISTOGRAM column before the COLUMN_STAT_HIST_SIZE. Please change the code in this function to do that)

spetrunia replied:
I think it is also fine to make `histogram_->serialize` return the number of bytes stored.

idoqo committed:
implement parse and serialize for histogram json

idoqo committed:
inform test result of zero hist_size for json histogram

Signed-off-by: Michael Okoko <okokomichaels@outlook.com>

idoqo committed:
report parse error when parsing JSON histogram fails

Signed-off-by: Michael Okoko <okokomichaels@outlook.com>

spetrunia committed:
MDEV-21130: Histograms: use JSON as on-disk format

A demo of how to use in-memory data structure for histogram.
The patch shows how to
* convert string form of data to binary form
* compare two values in binary form
* compute a fraction for val in [X, Y] range.

grep for GSOC-TODO for notes.

idoqo committed:
Use generic Histogram_base class for Histogram_builders

This fixes the wrong calculation for avg_frequency in json histograms
by replacing the specific histogram objects with the generic Histogram_base class.

It also restores get/set size functions as they were useful in calculating fields
for binary histogram.

Signed-off-by: Michael Okoko <okokomichaels@outlook.com>

idoqo committed:
fix returned value type for empty json objects

Signed-off-by: Michael Okoko <okokomichaels@outlook.com>
